### PR TITLE
[RLlib] Add examples and docs for Catalog

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

This PR adds examples and docs for RLlib's Catalog class which is used to create Models and distributions for RL Modules.

### CODE REVIEW

File: doc/source/_includes/rllib/rlmodules_rollout.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+..
+    .. admonition:: This doc is related to the RLModules API.
+
+        Disregard this doc if you do not explicitly want to use RLModules. The RLModules API is experimental and
```
Comment (avnishn) [inline doc/source/_includes/rllib/rlmodules_rollout.rst:4]: We could probably just say something to the effect of RLModules is an experimental API. I think this gets across the point overall. Its implied what experimental means.

Comment (ArturNiederfahrenhorst) [reply doc/source/_includes/rllib/rlmodules_rollout.rst:4]: Can you provide a complete sentence to replace this with?

Comment (avnishn) [reply doc/source/_includes/rllib/rlmodules_rollout.rst:4]: The RLModules api is currently in the experimental phase of its rollout.

Comment (ArturNiederfahrenhorst) [reply doc/source/_includes/rllib/rlmodules_rollout.rst:4]: Thanks!

Comment (avnishn) [reply doc/source/_includes/rllib/rlmodules_rollout.rst:4]: hmm while we're at it, is this a whole separate doc page right here? If so, why?

File: doc/source/rllib/rllib-catalogs.rst [hunk 2]
```diff
@@ -0,0 +1,151 @@
+.. include:: /_includes/rllib/rlmodules_rollout.rst
+
+.. _rllib-catalogs-user-guide:
+
+Catalogs
+========
+
+Catalogs are where RL Modules primarily get their models and action distributions from.
```
Comment (avnishn) [inline doc/source/rllib/rllib-catalogs.rst:8]: The `Catalog` is a tool that helps create default models and action distributions for RLModules to work with environments with any action and observation spaces. It provides an interface for constructing pre-made components like an MLP for policies that have continuous action spaces, and an action distribution (like a multivariate Gaussian) that matches that action space.

**Review by kouroshHakha — COMMENTED** [review]
I left several comments / suggestions. I had to overlook a lot of places that I would write / restructure things differently for the sake of time, but we can revisit them later, when reviewing our docs as a whole. 

One thing that applies to this PR in general is that there are so many places that you are referencing a class, but there is not cross reference to the class API. So I would suggest creating API reference for catalogs and cross referencing them where ever it is needed.  You can take a look at RLModule doc pr to see how many places I have referenced the api ref doc pages. This will immensely help with ease of navigation in your doc pages.

File: doc/source/_includes/rllib/rlmodules_rollout.rst (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply doc/source/_includes/rllib/rlmodules_rollout.rst]: It's an include. It's basically a banner that alerts users that they are looking at something experimental.
Angelina told me to make this basically the first thing people read - if it's not meant for John Doe, put it first.

CC @kouroshHakha to make you aware of this banner

File: doc/source/rllib/rllib-catalogs.rst [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+Prerequisites for this user guide is a rough understanding of RLModules.
+After reading this user guide you will be able to…
+
+    - Instantiate and interact with a Catalog
+    - Inject your custom models into RLModules
+    - Inject your custom action distributions into RLModules
+    - Extend RLlib’s selection of Models and distributions with your own
+    - Write a Catalog from scratch
+
+Instantiate and interact with a Catalog
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:41]: That's how I started out, @angelinalg wished for headings to reflect the list right above, such that people can go from that list right to the section they are interested in.

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: I have [a couple of issues](https://anyscaleteam.slack.com/archives/C02FL3S8DS6/p1680641091213999?thread_ts=1680641077.699009&cid=C02FL3S8DS6) with putting stuff in that folder.

I think only stuff like "RLLib in 60 seconds" belong there, stuff that's hyper-specific to docs.

It's very subjective what example code that we use in the docs constitutes an example that should also be under `rllib/examples` or goes under `doc/rllib/doc_code` - they are all examples of how to do stuff in RLlib at the end. 
Basically, stuff that no user would miss if they would scan our examples folder for examples should go in `rllib/doc_code`.
I don't see any point in putting what you could call "super small examples" in this folder.
Why not put all examples in the examples folder, instead of only putting end-to-end examples there?

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: TLDR is: I argue that the examples folder should also hold examples on basic usage of components because that (or docs page) are where our users look for examples. Since we don't want duplicates, they should simply be there because nobody ever scans the rllib/doc_code folder.

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: Wdyt?

File: doc/source/rllib/rllib-catalogs.rst [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    In this example, we showcase how to use the PPOCatalog to create models and an action distribution.
+    This is more similar to what RLlib does internally.
+
+    .. literalinclude:: ../../../rllib/examples/catalog/basics/cartpole_models_ppo.py
+       :language: python
+       :start-after: __sphinx_doc_begin__
+       :end-before: __sphinx_doc_end__
+
+.. tabbed:: Catalog + Custom Model + CartPole
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:65]: Mhhh. My idea was that "Basic interaction" is shown first and only if people click on these two others example tabs, they see this more advanced stuff.
We can also simply leave it out, I find it helpful to mentally connect Catalogs and Models though.

File: doc/source/rllib/rllib-catalogs.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    - Instantiate and interact with a Catalog
+    - Inject your custom models into RLModules
+    - Inject your custom action distributions into RLModules
+    - Extend RLlib’s selection of Models and distributions with your own
+    - Write a Catalog from scratch
+
+Instantiate and interact with a Catalog
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following three examples show variations of basic interaction with RLLib’s Catalogs.
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst]: Great idea, thanks!

File: doc/source/rllib/rllib-catalogs.rst [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    .. literalinclude:: ../../../rllib/examples/catalog/basics/cartpole_models.py
+       :language: python
+       :start-after: __sphinx_doc_begin__
+       :end-before: __sphinx_doc_end__
+
+What are Catalogs
+~~~~~~~~~~~~~~~~~
+
+Catalogs have two primary roles: Choosing the right model and choosing the right action distribution.
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:80]: I think we should prefer a simple style if it is not at the loss of precision.
@angelinalg wdyt?

File: doc/source/rllib/rllib-catalogs.rst [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    :align: center
+
+
+Inject your custom models into RLModules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can make Catalog build custom models by overriding the Catalog’s methods used by RL Modules to build models.
+Have a look at these lines from the constructor of the PPORLModules to see how Catalogs are being used by RLModules:
+
+.. literalinclude:: ../../../rllib/algorithms/ppo/ppo_base_rl_module.py
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:132]: Afaics, this collides with what Max said here: https://github.com/ray-project/ray/pull/33909#discussion_r1159317847
@maxpumperla do you have a rule here? Or am I misunderstanding on of you?

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:132]: Basically, I'm thinking about the risk that code outdates quickly.

File: doc/source/rllib/rllib-catalogs.rst [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+they are also part of RLlib’s configurations. As the primary entry point for configuring RLlib,
+AlgorithmConfig is the place where you can configure the Catalogs of the RLModules that are created.
+The following example shows how to configure the Catalogs of the RLModules that are created by PPO.
+
+.. literalinclude:: ../../../rllib/examples/catalog/basics/catalogs_in_algo_configs.py
+    :language: python
+    :start-after: __sphinx_doc_begin__
+    :end-before: __sphinx_doc_end__
+
+.. image:: images/catalog/catalog_rlmspecs_diagram.svg
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:122]: It's for illustration, not for extra clarification.
When skimming docs, the majority of information I ingest is through pictures.
Basically, users can get an idea of what they are looking at just by looking at the pictures to filter information without putting in lots of effort through reading.
@angelinalg lmk if this is stupid, I'm open to deleting this if I have the wrong idea here.

File: doc/source/rllib/rllib-catalogs.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:8]: Mhhh. I do this 5 lines below.
I don't see it being the first sentence in the RLModules docs either: https://docs.ray.io/en/master/rllib/rllib-rlmodule.html
.. or in other docs we have.
Why would it be different here?

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:8]: Maybe put it second or third? Like, the first sentence should be a very general, high-level description of what user is looking at I think.

File: doc/source/rllib/rllib-catalogs.rst [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    RLModules create an instance of the Catalog class they receive as part of their constructor.
+    They then create their internal models and action distributions with the help of this Catalog.
+
+    .. note::
+      You can also modify the models and distributions in RLModules directly by overriding their constructor!
+
+    .. image:: images/catalog/catalog_and_rlm_diagram.svg
+        :align: center
+
+.. tabbed:: PPOCatalog vs PPORLModule
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:97]: Got it.
What do you think about drop-downs?
I want to shield users from lenthy code sections that are not 100% relevant for what comes later in the doc.
Then, you can read through the doc with ease until you hit a section that you want to expand on because you have hit the depth that suits your use case.
For example here, you would not see the content of "PPOCatalog vs PPORLModule".

**Comment by ArturNiederfahrenhorst** [comment]
> One thing that applies to this PR in general is that there are so many places that you are referencing a class, but there is not cross reference to the class API. So I would suggest creating API reference for catalogs and cross referencing them where ever it is needed. You can take a look at RLModule doc pr to see how many places I have referenced the api ref doc pages. This will immensely help with ease of navigation in your doc pages.

100%! I also did that when I wrote other docs. I wanted to get your input on the text content though, just so that I don't have to change all the links afterwards 😄

File: doc/source/rllib/rllib-catalogs.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst:8]: Catalog is a secondary concept compared to RLModules and a lot of users (high-level API users) do not need to know about it unless they want to do some sort of extension to RLModules, that's why I think this is different. 

You can add it in a note format, similar to how the initial note is structured in RLModules. 
image

Something like 

Note: This abstraction is intended for advanced users who are interested in extending the functionality of build-in `~ray.rllib.core.rl_module.rl_module.RLModule` objects.

File: doc/source/rllib/rllib-catalogs.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst:121]: I think you need to add some extra caption to get the intention behind the the image across.

File: doc/source/rllib/rllib-catalogs.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst:131]: The context is a bit different. We want self-contained example code snippets to be tested all the time. But this is more like a high-lighted view of somewhere in the code-base. I think we should just snap that view out and put it in a self-contained context for the purpose of the doc.

File: doc/source/rllib/rllib-catalogs.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst:41]: I think that the list needs to also be updated to match this. So both should become `Basic Usage`. This is not in conflict with the point brought up by @angelinalg. The headings in my opinion should be short but also precise.

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: If we have good documentation with proper routing to the right examples, users won't have to scan the examples folder. rllib/doc_code is meant to be used for short and sweet copy and pastable snippets of code that go well with the flow of user-guides / tutorials presented in the docs. It's for internal house keeping and nobody is expected to go read the rllib/doc_code folder. They will also be auto-tested (You don't have to create new bazel files for them). For rllib/examples, I think these are more advanced cases, that are usually longer and are considered as "extra bonus" when it comes to documentation. That's why I think this distinction is important to be made here. Also for sake of keeping consistency with other libraries (like tune, data, air) we should do what I suggested.

[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: Just to be clear: 
I don't agree with this statement that 
```
we need to put all the examples in the rllib/examples folder because this is where users will look for all the examples. 
```
All users won't read our example folders to find things. Otherwise we wouldn't have so many discourse questions :)

File: doc/source/rllib/rllib-catalogs.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst]: I think drop down menu is a better UX for this in particular

File: doc/source/rllib/rllib-catalogs.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst:96]: I think drop down makes more sense. Also instead of PPOCatalog vs. PPORLModule you can title it as Example of catalog in PPORLModule

File: doc/source/rllib/rllib-catalogs.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:41]: Gotcha!

File: doc/source/rllib/rllib-catalogs.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst]: Thanks, changed.

File: doc/source/rllib/rllib-catalogs.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:131]: But the problem persists. We state here that the following is a snippet from the constructor of the PPORLModule.
If it is not, but rather a copy of it, the code will be outdated as soon as the constructor is changed.

File: doc/source/rllib/rllib-catalogs.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:121]: Gotcha, I've updated the illustration a little bit and changed the text around it to get the point across that in case of heterogeneous agents, you'll want to modify the multi-agent rl module spec.

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:121]: Thanks!

File: doc/source/rllib/rllib-catalogs.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:8]: Ok I agree.
Do you think we should add a similar matrix here?
Because afaics, 4 of the 5 columns here of at are actually Catalog features.

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:8]: I personally think it should be enough to have this matrix with RLModules.

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: I didn't say all users read the examples folder when looking for examples, but some do.
There are devs, like Sven for example, who simply don't read docs by default but gain much of their knowledge from reading code in a repo.
It's a headache for these users if we split up examples into end-to-end examples in ray/rllib/examples end "short and sweet copy and pastable snippets" that can be found in ray/doc/rllib/doc_code.

**Comment by ArturNiederfahrenhorst** [comment]
Sorry for the tags everyone, did a careless rebase 🙁

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: @kouroshHakha Also, there is a folder "rllib.examples.documentation".
With what you wrote above in mind, what would be the use of that one?

File: doc/source/rllib/rllib-catalogs.rst [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+.. literalinclude:: ../../../rllib/examples/catalog/basics/catalogs_in_algo_configs.py
+    :language: python
+    :start-after: __sphinx_doc_begin__
+    :end-before: __sphinx_doc_end__
+
+Basic Usage
+~~~~~~~~~~~
+
+The following three examples illustrate three basic usage patterns of Catalogs.
+The following example showcases the general API for interacting with Catalogs.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:77]:
```suggestion
The first example showcases the general API for interacting with Catalogs.
```

File: doc/source/rllib/rllib-catalogs.rst [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+.. literalinclude:: ../../../rllib/examples/catalog/basics/catalogs_in_algo_configs.py
+    :language: python
+    :start-after: __sphinx_doc_begin__
+    :end-before: __sphinx_doc_end__
+
+Basic Usage
+~~~~~~~~~~~
+
+The following three examples illustrate three basic usage patterns of Catalogs.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:76]:
```suggestion
The following three examples illustrate three basic usage patterns of Catalogs.

```

File: doc/source/rllib/rllib-catalogs.rst [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+The following example showcases the general API for interacting with Catalogs.
+
+.. literalinclude:: ../../../rllib/examples/catalog/basics/basic_interaction.py
+   :language: python
+   :start-after: __sphinx_doc_begin__
+   :end-before: __sphinx_doc_end__
+
+The following example showcases how to use the :py:class:`~ray.rllib.algorithms.ppo.ppo_catalog.PPOCatalog`
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:84]:
```suggestion
The second example showcases how to use the :py:class:`~ray.rllib.algorithms.ppo.ppo_catalog.PPOCatalog`
```

File: doc/source/rllib/rllib-catalogs.rst [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+.. dropdown:: **Use catalog-generated models**
+    :animate: fade-in-slide-down
+
+    .. literalinclude:: ../../../rllib/examples/catalog/basics/cartpole_models_ppo.py
+       :language: python
+       :start-after: __sphinx_doc_begin__
+       :end-before: __sphinx_doc_end__
+
+The following example showcases how to use the base :py:class:`~ray.rllib.core.models.catalog.Catalog`
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:96]:
```suggestion
The third example showcases how to use the base :py:class:`~ray.rllib.core.models.catalog.Catalog`
```

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: This is more of a consistency thing with other libraries. Documentation code should go to rllib/doc_code. So I think based on that we should also move rllib.examples.documentation to there in the longer term.

File: doc/source/rllib/rllib-catalogs.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply doc/source/rllib/rllib-catalogs.rst]: totally. I didn't mean to use this text, I meant to use the theme, i.e. the `..note:: ` syntax.

File: rllib/examples/catalog/basics/basic_interaction.py
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/examples/catalog/basics/basic_interaction.py:1]: @kouroshHakha Ok. tbh I still don't agree with you on putting the "basic examples" from these docs there instead of the examples folder, but I've followed your suggestion now to get this PR merged. It's not a dealbreaking difference for me anyway 🙂

**Review by kouroshHakha — APPROVED** [review]

File: doc/source/rllib/rllib-catalogs.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (angelinalg) [reply doc/source/rllib/rllib-catalogs.rst]: Yes, I tend to agree with simplicity, as long as you are not compromising technical accuracy.

File: doc/source/rllib/images/catalog/catalog_and_rlm_diagram.svg
Comment (angelinalg) [inline doc/source/rllib/images/catalog/catalog_and_rlm_diagram.svg:1]: The name of the object behind "action space" is difficult to read. I think it's the "observation space"?

Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/images/catalog/catalog_and_rlm_diagram.svg:1]: Yeah. It's not meant to be readable. It's a bunch of things that are described above that go into the Catalog.

Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/images/catalog/catalog_and_rlm_diagram.svg:1]: Should we change this?

File: doc/source/rllib/rllib-catalogs.rst [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+.. image:: images/catalog/catalog_and_rlm_diagram.svg
+    :align: center
+
+The following diagram shows a concrete case in more detail.
+
+.. dropdown:: **Example of catalog in a PPORLModule**
+    :animate: fade-in-slide-down
+
+    The :py:class:`~ray.rllib.algorithms.ppo.ppo_catalog.PPOCatalog` is fed an `observation space``, ``action space``,
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:134]:
```suggestion
    The :py:class:`~ray.rllib.algorithms.ppo.ppo_catalog.PPOCatalog` is fed an ``observation space``, ``action space``,
```

File: doc/source/rllib/rllib-catalogs.rst [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+    a ``model config dict`` and the ``view requirements`` of the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`.
+    The model config dicts and the view requirements are only of interest in special cases, such as
+    recurrent networks or attention networks. A PPORLModule has four components that are created by the
+    :py:class:`~ray.rllib.algorithms.ppo.ppo_catalog.PPOCatalog`:
+    ``Encoder``, ``value function head``, ``policy head`` and ``action distribution``. You can find out more about this
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:139]:
```suggestion
    ``Encoder``, ``value function head``, ``policy head``, and ``action distribution``. You can find out more about this
```

Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:139]:
```suggestion
    ``Encoder``, ``value function head``, ``policy head`` and ``action distribution``. Find out more about the
```

File: doc/source/rllib/rllib-catalogs.rst [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    distinction between these components in our section on Models.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:140]: Do you want to link to the section on Models?

File: doc/source/rllib/package_ref/catalogs.rst [hunk 17]
```diff
@@ -0,0 +1,41 @@
+
+
+.. _catalog-reference-docs:
+
+.. include:: /_includes/rllib/rlmodules_rollout.rst
+
+Catalog API
+===========
+
+Basic Usage
```
Comment (angelinalg) [inline doc/source/rllib/package_ref/catalogs.rst:10]:
```suggestion
Basic usage
```

File: doc/source/rllib/rllib-catalogs.rst (hunk 16 above)
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:140]: I'm getting rid of this sentence and will add this info back once the section is added...

File: doc/source/rllib/package_ref/catalogs.rst [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+-----------
+
+The follwing basic API can be used to get a default ``encoder`` or ``action distribution``
+out of Catalog. You can inherit from Catalog and modify the following methods to
+directly inject custom components into a given RLModule.
+Algorithm-specific implementations of Catalog have additional methods.
+For example for building ``heads``.
```
Comment (angelinalg) [inline doc/source/rllib/package_ref/catalogs.rst:17]:
```suggestion
for example, for building ``heads``.
```

File: doc/source/rllib/package_ref/catalogs.rst [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+.. autosummary::
+    :toctree: doc/
+
+    Catalog
+    Catalog.build_encoder
+    Catalog.get_action_dist_cls
+    Catalog.get_preprocessor
+
+
+Advanced Usage
```
Comment (angelinalg) [inline doc/source/rllib/package_ref/catalogs.rst:30]:
```suggestion
Advanced usage
```

File: doc/source/rllib/rllib-catalogs.rst [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+- How to inject a custom action distribution into a :py:class:`~ray.rllib.core.models.catalog.Catalog`
+
+.. literalinclude:: ../../../rllib/examples/catalog/custom_action_distribution.py
+   :language: python
+   :start-after: __sphinx_doc_begin__
+   :end-before: __sphinx_doc_end__
+
+
+
+Notable TODOs
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:173]: Do you want to hide this list?

File: doc/source/rllib/rllib-catalogs.rst [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+    distinction between these components in our section on Models.
+
+    .. image:: images/catalog/ppo_catalog_and_rlm_diagram.svg
+        :align: center
+
+
+Inject your custom models into RLModules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can make a :py:class:`~ray.rllib.core.models.catalog.Catalog` build custom ``models`` by overriding the Catalog’s methods used by RL Modules to build ``models``.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-catalogs.rst:148]:
```suggestion
You can make a :py:class:`~ray.rllib.core.models.catalog.Catalog` build custom ``models`` by overriding the Catalog’s methods used by RLModules to build ``models``.
```

**Comment by angelinalg** [comment]
There is a mix of using `RL Module` and `RLModule` in this file: [doc/source/rllib/rllib-rlmodule.rst](https://github.com/ray-project/ray/pull/33898/files#diff-92905f359d6a3911301383f4549fdeec901505b2c25be4bdb81ad7742cbeca42). I suggest being consistent. The majority of occurrences seem to be `RLModule`. The title should probably be consistent.

File: doc/source/rllib/rllib-catalogs.rst (hunk 20 above)
Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:173]: It's not being hidden in other docs we have, but I think it's actually a good idea to hide it.

Comment (ArturNiederfahrenhorst) [reply doc/source/rllib/rllib-catalogs.rst:173]: CC @kouroshHakha

**Comment by angelinalg** [comment]
Sorry about closing that by mistake, @ArturNiederfahrenhorst

File: doc/source/rllib/rllib-rlmodule.rst [hunk 22]
```diff
@@ -263,9 +263,9 @@ There are two possible ways to extend existing RLModules:
     
 .. tabbed:: Extending RLModule Catalog
 
-    Another way to customize your module is by extending its catalog. The catalog is a component that defines the default architecture and behavior of a model based on factors such as ``observation_space``, ``action_space``, etc. To modify sub-components of an existing RLModule, extend the corresponding catalog class.
+    Another way to customize your module is by extending its :py:class:`~ray.rllib.core.models.catalog.Catalog`. The :py:class:`~ray.rllib.core.models.catalog.Catalog` is a component that defines the default architecture and behavior of a model based on factors such as ``observation_space``, ``action_space``, etc. To modify sub-components of an existing RLModule, extend the corresponding catalog class.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:266]:
```suggestion
    Another way to customize your module is by extending its :py:class:`~ray.rllib.core.models.catalog.Catalog`. The :py:class:`~ray.rllib.core.models.catalog.Catalog` is a component that defines the default architecture and behavior of a model based on factors such as ``observation_space``, ``action_space``, etc. To modify sub-components of an existing RLModule, extend the corresponding Catalog class.
```

Comment (angelinalg) [reply doc/source/rllib/rllib-rlmodule.rst:266]: I suspect `Catalog` is a special class, which is why I'm suggesting to capitalize it. Please ignore if I'm mistaken.

**Comment by ArturNiederfahrenhorst** [comment]
> There is a mix of using `RL Module` and `RLModule` in this file: [doc/source/rllib/rllib-rlmodule.rst](https://github.com/ray-project/ray/pull/33898/files#diff-92905f359d6a3911301383f4549fdeec901505b2c25be4bdb81ad7742cbeca42). I suggest being consistent. The majority of occurrences seem to be `RLModule`. The title should probably be consistent.

@kouroshHakha I'm making this change requested by Angeline within the scope of this PR.
Lmk if you object.

File: doc/source/rllib/user-guides.rst [hunk 23]
```diff
@@ -98,3 +98,12 @@ RLlib Feature Guides
...
         :text: How To Work With the RLlib CLI?
         :classes: btn-link btn-block stretched-link
+
+    ---
+    :img-top: /rllib/images/rllib-logo.svg
+
+    +++
+    .. link-button:: rllib-catalogs
+        :type: ref
+        :text: How To Use the RLlib Catalogs?
```
Comment (angelinalg) [inline doc/source/rllib/user-guides.rst:108]:
```suggestion
        :text: How To Use the RLlib Catalogs
```

Comment (angelinalg) [reply doc/source/rllib/user-guides.rst:108]: The construction, "How to Use X" doesn't need a question mark. But I can see the the other cards use question marks, too. If you have time to fix the other occurrences, please do.

File: rllib/examples/catalog/custom_action_distribution.py [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+    def kl(self, other):
+        return torch.distributions.kl.kl_divergence(self.torch_dist, other.torch_dist)
+
+    @staticmethod
+    def required_input_dim(space):
+        return int(space.n)
+
+    @classmethod
+    # This method is used to create distributions from logits inside RLModules
```
Comment (angelinalg) [inline rllib/examples/catalog/custom_action_distribution.py:42]:
```suggestion
    # This method is used to create distributions from logits inside RLModules.
```

**Comment by angelinalg** [comment]
@ArturNiederfahrenhorst, I did a quick pass and left some minor comments. I really like your diagrams!

**Comment by ArturNiederfahrenhorst** [comment]
> @ArturNiederfahrenhorst, I did a quick pass and left some minor comments. I really like your diagrams!

Awesome, thanks you so much!

File: doc/source/rllib/rllib-catalogs.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (angelinalg) [reply doc/source/rllib/rllib-catalogs.rst]: Generally it's good to have diagrams for visual learners.

**Review by angelinalg — APPROVED** [review]
LGTM

**Review by maxpumperla — APPROVED** [review]
